src/v/cluster/controller.cc

```diff
@@ -576,31 +626,15 @@ ss::future<> controller::cluster_creation_hook(cluster_discovery& discovery) {
             co_await ss::sleep(100ms);
         }
 
-        co_return co_await create_cluster();
+        co_return co_await create_cluster(create_cluster_mode::bootstrap, [] {
+            return ss::make_ready_future<ss::stop_iteration>(
+              ss::stop_iteration::no);
+        });
     }
 
-    // The new cluster creation is done by the leader of raft0 consisting of
-    // seed servers, once elected
-    while (!_storage.local().get_cluster_uuid() && !_raft0->is_leader()) {
-        co_await ss::sleep(100ms);
-    }
-
-    if (_storage.local().get_cluster_uuid()) {
-        vlog(
-          clusterlog.info,
-          "Cluster is {}",
-          *_storage.local().get_cluster_uuid());
-        co_return;
-    }
-
-    if (
-      _raft0->last_visible_index() > model::offset{}
-      || _raft0->config().brokers().size() > 1) {
-        // The controller log has already been written to
-        // TODO: create_cluster() but use the metrics' cluster_id
-    }
-
-    co_return co_await create_cluster();
+    co_return co_await create_cluster(create_cluster_mode::bootstrap, [this] {
+        return seeds_driven_leader_barrier();
+    });
```

Comment:
Isn't it harmless to call `seeds_driven_leader_barrier()` for root-driven mode? IMHO we should just inline the barrier unconditionally to `create_cluster()` and just eat the extra cycles we're scheduling in root-driven mode. I think long term the current code structure will be difficult to follow and maintain.

Reply:
Well it is harmless but the needless checking for cluster uuid would be confusing for understanding the intentions. On the other hand, layering this check out into a strategy is not the clearest way either.

Reply:
I actually think it'd be less confusing to do the needless check, because the check itself is intuitive: "wait until there's a UUID or we are leader". That is the intention of that block. It's only a special case / optimization for a single-node deployments because they don't _have_ to wait for leader because we assume that we are the leader.

Reply:
ok